enterprise/backend/src/metabase_enterprise/scim/v2/api.clj

```diff
@@ -1,14 +1,419 @@
+          membership->group    (fn [membership] (select-keys membership [:name :entity_id]))]
+      (for [user users]
+        (assoc user :user_group_memberships (->> (user-id->memberships (u/the-id user))
+                                                 (map membership->group)
+                                                 (sort-by :entity_id)))))))
+
+(mu/defn ^:private mb-user->scim :- SCIMUser
+  "Given a Metabase user, returns a SCIM user."
+  [user]
+  {:schemas  [user-schema-uri]
+   :id       (:entity_id user)
+   :userName (:email user)
+   :name     {:givenName  (:first_name user)
+              :familyName (:last_name user)}
+   :emails   [{:value (:email user)
+               :type    "work"
+               :primary true}]
+   :groups   (map
+              (fn [membership]
+                {:value   (:entity_id membership)
+                 :$ref    (str (scim/scim-base-url) "/Groups/" (:entity_id membership))
+                 :display (:name membership)})
+              (:user_group_memberships user))
+   :locale   (:locale user)
+   :active   (:is_active user)
+   :meta     {:resourceType "User"}})
+
+(mu/defn ^:private scim-user->mb :- user/NewUser
+  "Given a SCIM user, returns a Metabase user."
+  [user]
+  (let [{email :userName name-obj :name locale :locale is-active? :active} user
+        {:keys [givenName familyName]} name-obj]
+    (merge
+     {:first_name givenName
+      :last_name  familyName
+      :email      email
+      :is_active  is-active?
+      :type       :personal}
+     (when (and locale (i18n/available-locale? locale))
+       {:locale locale}))))
+
+(mu/defn ^:private get-user-by-entity-id
+  "Fetches a user by entity ID, or throws a 404"
+  [entity-id]
+  (or (t2/select-one (cons :model/User user-cols)
+                     :entity_id entity-id
+                     {:where [:= :type "personal"]})
+      (throw-scim-error 404 "User not found")))
+
+(defn- ^:private user-filter-clause
+  [filter-parameter]
+  (let [[_ match] (re-matches #"^userName eq \"(.*)\"$" filter-parameter)]
+    (if match
+      [:= :%lower.email (u/lower-case-en match)]
+      (throw-scim-error 400 (format "Unsupported filter parameter: %s" filter-parameter)))))
+
 (defendpoint GET "/Users"
-  "Fetches a list of users."
-  []
-  (t2/select :model/User))
+  "Fetch a list of users."
+  [:as {{start-index :startIndex c :count filter-param :filter} :params}]
+  {start-index  [:maybe ms/PositiveInt]
+   c            [:maybe ms/PositiveInt]
+   filter-param [:maybe ms/NonBlankString]}
+  (let [limit          (or c default-pagination-limit)
+        ;; SCIM start-index is 1-indexed, so we need to decrement it here
+        offset         (if start-index (dec start-index) default-pagination-offset)
+        filter-param   (when filter-param (codec/url-decode filter-param))
+        where-clause   [:and [:= :type "personal"]
+                             (when filter-param (user-filter-clause (codec/url-decode filter-param)))]
```

Comment:
Looks like we're url-decoding `filter-param` multiple times here? Once in the `let` and then again here.